### PR TITLE
[SPARK-5642] [SQL] Apply column pruning on unused aggregation fields

### PR DESCRIPTION
**Pull request by adrian-wang**
select k from (select key k, max(value) v from src group by k) t

### CODE REVIEW

**Comment by rxin** [comment]
Is it possible to add a unit test?

**Comment by rxin** [comment]
(I understand unit test coverage for the optimizer is pretty low - but that would be great to change & increase)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -116,6 +116,10 @@ object ColumnPruning extends Rule[LogicalPlan] {
     case a @ Aggregate(_, _, child) if (child.outputSet -- a.references).nonEmpty =>
       a.copy(child = Project(a.references.toSeq, child))
 
+    case p @ Project(pl, Aggregate(ge, ec, child))
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I'd follow the convention from the rest of the optimizer and use longer variable names (i.e. `projectList`, `groupExprs`, `aggExprs`)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -116,6 +116,10 @@ object ColumnPruning extends Rule[LogicalPlan] {
...
+      if (ec.map(_.references).toSet -- p.references).nonEmpty =>
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: indent 4.

Also I don't think this is doing what you want.  `ec.map(_.references).toSet` return a `Set[AttributeSet]`.

Instead you probably want `(a.outputSet -- p.references).nonEmpty` where `a` is bound to the `Aggregate`

**Comment by adrian-wang** [comment]
@rxin @marmbrus

**Comment by marmbrus** [comment]
Thanks!  I'm going to go ahead and add one more test case while merging since I wasn't sure if this case was handled properly (it is :) )

Merging to master and 1.3

**Comment by marmbrus** [comment]
```scala
  test("column pruning for group with alias") {
    val originalQuery =
      testRelation
        .groupBy('a)('a as 'c, Count('b))
        .select('c)

    val optimized = Optimize(originalQuery.analyze)
    val correctAnswer =
      testRelation
        .select('a)
        .groupBy('a)('a as 'c)
        .select('c).analyze

    comparePlans(optimized, correctAnswer)
  }
```

**Pull request closed without merging**